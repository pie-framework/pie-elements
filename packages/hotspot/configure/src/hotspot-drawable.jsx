import React from 'react';
import PropTypes from 'prop-types';
import { Layer, Stage } from 'react-konva';
import max from 'lodash/max';
import { withStyles } from '@material-ui/core/styles/index';

import Rectangle from './hotspot-rectangle';
import Polygon from './hotspot-polygon';

export class Drawable extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isDrawing: false,
      stateShapes: false,
      dimensions: {
        height: 0,
        width: 0,
      }
    };
  }

  handleOnStageClick = (e) => {
    // !IMPORTANT: currently, only 'rectangles' are drawable
    const { isDrawing, stateShapes } = this.state;
    const { onUpdateShapes } = this.props;

    // If we're drawing a shape, a click finishes the drawing and sends the new shapes to HOC
    if (isDrawing) {
      this.setState({ isDrawing: !isDrawing, stateShapes: false, isDrawingShapeId: undefined });

      if (stateShapes) {
        onUpdateShapes(stateShapes);
      }

      return;
    }

    // Otherwise, add a new rectangle at the mouse position with 0 width and height
    const newShapes = this.props.shapes.slice();
    const value = max(newShapes.map(c => parseInt(c.id)).filter(id => !isNaN(id))) || 0;

    newShapes.push({
      id: `${value + 1}`,
      height: 0,
      width: 0,
      x: e.evt.layerX,
      y: e.evt.layerY,
      type: 'rectangles'
    });

    this.setState({ isDrawing: true, isDrawingShapeId: `${value + 1}` });
    onUpdateShapes(newShapes);
  };

  handleMouseMove = (e) => {
    // !IMPORTANT: currently, only 'rectangles' are drawable
    const { isDrawing, isDrawingShapeId } = this.state;
    const { shapes } = this.props;

    if (isDrawing && isDrawingShapeId) {
      const mouseX = e.evt.layerX;
      const mouseY = e.evt.layerY;

      const currShapeIndex = shapes.findIndex(shape => shape.id === isDrawingShapeId);
      const currShape = shapes[currShapeIndex];

      if (currShape) {
        const newWidth = mouseX - currShape.x;
        const newHeight = mouseY - currShape.y;

        const newShapesList = shapes.slice();

        newShapesList[currShapeIndex] = {
          ...newShapesList[currShapeIndex],
          height: newHeight,
          width: newWidth,
          x: currShape.x,
          y: currShape.y
        };

        // On mouse move don't trigger any event. Put the shapes on this state instead.
        this.setState({ stateShapes: newShapesList });
      }
    }
  };

  handleOnSetAsCorrect = (shape) => {
    const { id } = shape;
    const { multipleCorrect, shapes, onUpdateShapes } = this.props;

    let newShapes = [];

    if (multipleCorrect) {
      newShapes = shapes.map(shape => {
        if (shape.id === id) {
          shape.correct = !shape.correct;
        }
        return shape;
      })
    } else {
      newShapes = shapes.map(shape => {
        shape.correct = shape.id === id;
        return shape;
      })
    }

    onUpdateShapes(newShapes);
  };

  handleOnDragEnd = (id, updatedProps) => {
    const { shapes, onUpdateShapes } = this.props;
    const newShapes = shapes.map(shape => {
      if (shape.id === id) {
        return {
          ...shape,
          ...updatedProps
        }
      }
      return shape;
    });

    onUpdateShapes(newShapes);
  };

  handleOnImageLoad = ({ target }) => {
    const { onUpdateImageDimension } = this.props;
    const resizeHandle = this.resize;
    const elementStyle = getComputedStyle(target);
    const newHeight = parseFloat(elementStyle.height);
    const newWidth = parseFloat(elementStyle.width);

    const dimensions = {
      height: newHeight,
      width: newWidth
    };

    this.setState({ dimensions });
    onUpdateImageDimension(dimensions);

    resizeHandle.addEventListener('mousedown', this.initialiseResize, false);
  };

  initialiseResize = () => {
    window.addEventListener('mousemove', this.startResizing, false);
    window.addEventListener('mouseup', this.stopResizing, false);
  };

  startResizing = (e) => {
    const box = this.image;
    const { disableDrag } = this.props;

    const bounds = e.target.getBoundingClientRect();
    const x = e.clientX - bounds.left;
    const y = e.clientY - bounds.top;

    const resizeValid = this.checkIfResizeValid(x, y);
    const hasMinimumWidth = x > 150 && y > 150;

    if (resizeValid && hasMinimumWidth) {
      box.style.width = `${x}px`;
      box.style.height = `${y}px`;

      this.setState({ dimensions: { height: y, width: x } });
    }

    disableDrag();
  };

  checkIfResizeValid(x, y) {
    const { shapes } = this.state;
    let drawable = true;

    // Do not allow resizing over the hotspots
    shapes && shapes.forEach(shape => {
      const right = shape.x + shape.width + 5;
      const bottom = shape.y + shape.height + 5;
      if (x <= right || y <= bottom) {
        drawable = false
      }
    });
    return drawable;
  }

  stopResizing = () => {
    const { enableDrag, onUpdateImageDimension } = this.props;
    const { dimensions } = this.state;

    enableDrag();
    onUpdateImageDimension(dimensions);

    window.removeEventListener('mousemove', this.startResizing, false);
    window.removeEventListener('mouseup', this.stopResizing, false);
  };

  render() {
    const {
      classes,
      imageUrl,
      dimensions: { height, width },
      hotspotColor,
      outlineColor,
      shapes,
      strokeWidth
    } = this.props;
    const {
      stateShapes,
      isDrawing,
      dimensions: { height: heightFromState, width: widthFromState }
    } = this.state;
    const shapesToUse = stateShapes || shapes;

    return (
      <div className={classes.base}>
        {imageUrl && (
          <div className={classes.imageContainer}>
            <img
              className={classes.image}
              onLoad={this.handleOnImageLoad}
              ref={ref => {
                this.image = ref;
              }}
              src={imageUrl}
              {...height && width ? { style: { height, width } } : {}}
            />
            <div ref={ref => {
              this.resize = ref;
            }} className={classes.resize}/>
          </div>
        )}

        <Stage
          className={classes.stage}
          height={heightFromState || height}
          onClick={this.handleOnStageClick}
          onContentMouseMove={this.handleMouseMove}
          width={widthFromState || width}
        >
          <Layer>
            {shapesToUse.map((shape, index) => {
              const Tag = shape.type === 'polygons' ? Polygon : Rectangle;

              return (
                <Tag
                  correct={shape.correct}
                  isDrawing={isDrawing}
                  height={shape.height}
                  hotspotColor={hotspotColor}
                  id={shape.id}
                  key={index}
                  onClick={() => this.handleOnSetAsCorrect(shape)}
                  onDragEnd={this.handleOnDragEnd}
                  outlineColor={outlineColor}
                  width={shape.width}
                  x={shape.x}
                  y={shape.y}
                  points={shape.points}
                  strokeWidth={strokeWidth}
                />
              );
            })}
          </Layer>
        </Stage>
      </div>
    );
  }
}

const styles = () => ({
  base: {
    position: 'relative'
  },
  image: {
    alignItems: 'center',
    display: 'flex',
    justifyContent: 'center',
  },
  imageContainer: {
    position: 'relative',
    width: 'fit-content'
  },
  resize: {
    borderBottom: '1px solid #727272',
    borderRight: '1px solid #727272',
    bottom: '-10px',
    cursor: 'se-resize',
    height: '10px',
    position: 'absolute',
    right: '-10px',
    width: '10px'
  },
  stage: {
    left: 0,
    top: 0,
    position: 'absolute'
  }
});

Drawable.propTypes = {
  classes: PropTypes.object.isRequired,
  disableDrag: PropTypes.func.isRequired,
  dimensions: PropTypes.object.isRequired,
  enableDrag: PropTypes.func.isRequired,
  imageUrl: PropTypes.string.isRequired,
  hotspotColor: PropTypes.string.isRequired,
  multipleCorrect: PropTypes.bool.isRequired,
  onUpdateImageDimension: PropTypes.func.isRequired,
  onUpdateShapes: PropTypes.func.isRequired,
  outlineColor: PropTypes.string.isRequired,
  shapes: PropTypes.array.isRequired,
  strokeWidth: PropTypes.number
};

export default withStyles(styles)(Drawable);
