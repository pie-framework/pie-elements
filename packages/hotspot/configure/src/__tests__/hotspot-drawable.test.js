import { shallow } from 'enzyme';
import React from 'react';
import { Drawable } from '../hotspot-drawable';

const model = () => ({
  imageUrl: 'https://cdn.fluence.net/image/0240eb1455ce4c4bb6180232347b6aef_W',
  shapes: [
    {
      id: '0',
      height: 140,
      width: 130,
      x: 1,
      y: 1,
      correct: true,
      type: 'rectangles'
    }, {
      id: '1',
      height: 140,
      width: 130,
      x: 140,
      y: 1,
      type: 'rectangles'
    }, {
      id: '2',
      height: 140,
      width: 130,
      x: 280,
      y: 1,
      type: 'rectangles'
    },
    {
      id: '3',
      points: [
        { x: 1, y: 148 },
        { x: 1, y: 288 },
        { y: 288, x: 129 },
        { y: 148, x: 129 }],
      correct: true,
      type: 'polygons'
    }, {
      id: '4',
      points: [
        { y: 151, x: 141 },
        { y: 289, x: 141 },
        { y: 289, x: 269 },
        { x: 269, y: 151 }],
      correct: false,
      type: 'polygons'
    }, {
      id: '5',
      points: [
        { x: 279, y: 150 },
        { x: 279, y: 289 },
        { x: 407, y: 289 },
        { x: 407, y: 150 }
      ],
      correct: false,
      type: 'polygons'
    }
  ],
  dimensions: {
    height: 291, width: 410
  },
  hotspotColor: 'rgba(137, 183, 244, 0.65)',
  outlineColor: 'blue',
  multipleCorrect: true
});

describe('HotspotDrawable', () => {
  let w,
    handleDisableDrag = jest.fn(),
    handleEnableDrag = jest.fn(),
    onUpdateImageDimension = jest.fn(),
    onUpdateShapes = jest.fn(),
    initialModel = model()
  ;

  beforeEach(() => {
    w = extras => {
      const props = {
        classes: {},
        dimensions: initialModel.dimensions,
        disableDrag: handleDisableDrag,
        enableDrag: handleEnableDrag,
        imageUrl: initialModel.imageUrl,
        hotspotColor: initialModel.hotspotColor,
        multipleCorrect: initialModel.multipleCorrect,
        onUpdateImageDimension: onUpdateImageDimension,
        onUpdateShapes: onUpdateShapes,
        outlineColor: initialModel.outlineColor,
        shapes: initialModel.shapes,
        ...extras
      };

      return shallow(<Drawable {...props} />);
    }
  });

  describe('render', () => {
    it('renders with default strokeWidth', () => {
      expect(w()).toMatchSnapshot();
    });

    it('renders with given strokeWidth', () => {
      expect(w({ strokeWidth: 10 })).toMatchSnapshot();
    });

    it('renders', () => {
      expect(w()).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = w();
    });

    it('handleOnStageClick isDrawing = false', () => {
      wrapper.instance().handleOnStageClick({
        evt: {
          layerX: 20,
          layerY: 30
        }
      });

      expect(onUpdateShapes).toHaveBeenCalledWith([
        ...initialModel.shapes,
        {
          id: '6',
          height: 0,
          width: 0,
          x: 20,
          y: 30,
          type: 'rectangles'
        }
      ]);
    });

    it('handleOnStageClick isDrawing = true', () => {
      wrapper.instance().state.isDrawing = true;
      wrapper.instance().state.stateShapes = initialModel.shapes.slice(0, 2);

      wrapper.instance().handleOnStageClick({
        evt: {
          layerX: 20,
          layerY: 30
        }
      });

      expect(onUpdateShapes).toHaveBeenCalledWith([...initialModel.shapes.slice(0, 2)]);

      // at this point, state.stateShapes is false, so we don't want to update shapes with false (onUpdateShapes)
      expect(wrapper.instance().state.stateShapes).toEqual(false);

      wrapper.instance().handleOnStageClick({
        evt: {
          layerX: 20,
          layerY: 30
        }
      });

      expect(onUpdateShapes).not.toBeCalledWith(false);
    });

    it('handleOnStageClick isDrawing = true', () => {
      wrapper.instance().state.isDrawing = true;
      wrapper.instance().state.stateShapes = initialModel.shapes.slice(0, 2);

      wrapper.instance().handleOnStageClick({});

      expect(onUpdateShapes).toHaveBeenCalledWith(initialModel.shapes.slice(0, 2))
    });

    it('handleOnSetAsCorrect correct', () => {
      wrapper.instance().handleOnSetAsCorrect(
        {
          id: '1'
        }
      );

      expect(onUpdateShapes).toHaveBeenCalledWith([
        initialModel.shapes[0],
        {
          ...initialModel.shapes[1],
          correct: true
        },
        ...initialModel.shapes.slice(2)
      ]);
    });

    it('handleOnSetAsCorrect incorrect', () => {
      wrapper.instance().handleOnSetAsCorrect(
        {
          id: '0'
        }
      );

      expect(onUpdateShapes).toHaveBeenCalledWith([
        {
          ...initialModel.shapes[0],
          correct: false
        },
        ...initialModel.shapes.slice(1)
      ]);
    });

    it('handleOnDragEnd', () => {
      wrapper.instance().handleOnDragEnd('0', { x: 1, y: 1 });

      expect(onUpdateShapes).toHaveBeenCalledWith([
        {
          ...initialModel.shapes[0],
          x: 1,
          y: 1
        },
        ...initialModel.shapes.slice(1)
      ]);
    });

    it('handleOnDragEnd unexistent id', () => {
      wrapper.instance().handleOnDragEnd('10', { x: 1, y: 1 });

      expect(onUpdateShapes).toHaveBeenCalledWith(initialModel.shapes);
    });

  });
});
