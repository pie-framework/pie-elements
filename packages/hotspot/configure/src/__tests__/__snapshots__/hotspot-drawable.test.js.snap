// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotspotDrawable render renders 1`] = `
<div>
  <div>
    <img
      onLoad={[Function]}
      src="https://cdn.fluence.net/image/0240eb1455ce4c4bb6180232347b6aef_W"
      style={
        Object {
          "height": 291,
          "width": 410,
        }
      }
    />
    <div />
  </div>
  <ForwardRef
    height={291}
    onClick={[Function]}
    onContentMouseMove={[Function]}
    width={410}
  >
    <Layer>
      <WithStyles(RectComponent)
        correct={true}
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="0"
        isDrawing={false}
        key="0"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        width={130}
        x={1}
        y={1}
      />
      <WithStyles(RectComponent)
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="1"
        isDrawing={false}
        key="1"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        width={130}
        x={140}
        y={1}
      />
      <WithStyles(RectComponent)
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="2"
        isDrawing={false}
        key="2"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        width={130}
        x={280}
        y={1}
      />
      <WithStyles(PolComponent)
        correct={true}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="3"
        isDrawing={false}
        key="3"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 1,
              "y": 148,
            },
            Object {
              "x": 1,
              "y": 288,
            },
            Object {
              "x": 129,
              "y": 288,
            },
            Object {
              "x": 129,
              "y": 148,
            },
          ]
        }
      />
      <WithStyles(PolComponent)
        correct={false}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="4"
        isDrawing={false}
        key="4"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 141,
              "y": 151,
            },
            Object {
              "x": 141,
              "y": 289,
            },
            Object {
              "x": 269,
              "y": 289,
            },
            Object {
              "x": 269,
              "y": 151,
            },
          ]
        }
      />
      <WithStyles(PolComponent)
        correct={false}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="5"
        isDrawing={false}
        key="5"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 279,
              "y": 150,
            },
            Object {
              "x": 279,
              "y": 289,
            },
            Object {
              "x": 407,
              "y": 289,
            },
            Object {
              "x": 407,
              "y": 150,
            },
          ]
        }
      />
    </Layer>
  </ForwardRef>
</div>
`;

exports[`HotspotDrawable render renders with default strokeWidth 1`] = `
<div>
  <div>
    <img
      onLoad={[Function]}
      src="https://cdn.fluence.net/image/0240eb1455ce4c4bb6180232347b6aef_W"
      style={
        Object {
          "height": 291,
          "width": 410,
        }
      }
    />
    <div />
  </div>
  <ForwardRef
    height={291}
    onClick={[Function]}
    onContentMouseMove={[Function]}
    width={410}
  >
    <Layer>
      <WithStyles(RectComponent)
        correct={true}
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="0"
        isDrawing={false}
        key="0"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        width={130}
        x={1}
        y={1}
      />
      <WithStyles(RectComponent)
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="1"
        isDrawing={false}
        key="1"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        width={130}
        x={140}
        y={1}
      />
      <WithStyles(RectComponent)
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="2"
        isDrawing={false}
        key="2"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        width={130}
        x={280}
        y={1}
      />
      <WithStyles(PolComponent)
        correct={true}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="3"
        isDrawing={false}
        key="3"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 1,
              "y": 148,
            },
            Object {
              "x": 1,
              "y": 288,
            },
            Object {
              "x": 129,
              "y": 288,
            },
            Object {
              "x": 129,
              "y": 148,
            },
          ]
        }
      />
      <WithStyles(PolComponent)
        correct={false}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="4"
        isDrawing={false}
        key="4"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 141,
              "y": 151,
            },
            Object {
              "x": 141,
              "y": 289,
            },
            Object {
              "x": 269,
              "y": 289,
            },
            Object {
              "x": 269,
              "y": 151,
            },
          ]
        }
      />
      <WithStyles(PolComponent)
        correct={false}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="5"
        isDrawing={false}
        key="5"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 279,
              "y": 150,
            },
            Object {
              "x": 279,
              "y": 289,
            },
            Object {
              "x": 407,
              "y": 289,
            },
            Object {
              "x": 407,
              "y": 150,
            },
          ]
        }
      />
    </Layer>
  </ForwardRef>
</div>
`;

exports[`HotspotDrawable render renders with given strokeWidth 1`] = `
<div>
  <div>
    <img
      onLoad={[Function]}
      src="https://cdn.fluence.net/image/0240eb1455ce4c4bb6180232347b6aef_W"
      style={
        Object {
          "height": 291,
          "width": 410,
        }
      }
    />
    <div />
  </div>
  <ForwardRef
    height={291}
    onClick={[Function]}
    onContentMouseMove={[Function]}
    width={410}
  >
    <Layer>
      <WithStyles(RectComponent)
        correct={true}
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="0"
        isDrawing={false}
        key="0"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        strokeWidth={10}
        width={130}
        x={1}
        y={1}
      />
      <WithStyles(RectComponent)
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="1"
        isDrawing={false}
        key="1"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        strokeWidth={10}
        width={130}
        x={140}
        y={1}
      />
      <WithStyles(RectComponent)
        height={140}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="2"
        isDrawing={false}
        key="2"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        strokeWidth={10}
        width={130}
        x={280}
        y={1}
      />
      <WithStyles(PolComponent)
        correct={true}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="3"
        isDrawing={false}
        key="3"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 1,
              "y": 148,
            },
            Object {
              "x": 1,
              "y": 288,
            },
            Object {
              "x": 129,
              "y": 288,
            },
            Object {
              "x": 129,
              "y": 148,
            },
          ]
        }
        strokeWidth={10}
      />
      <WithStyles(PolComponent)
        correct={false}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="4"
        isDrawing={false}
        key="4"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 141,
              "y": 151,
            },
            Object {
              "x": 141,
              "y": 289,
            },
            Object {
              "x": 269,
              "y": 289,
            },
            Object {
              "x": 269,
              "y": 151,
            },
          ]
        }
        strokeWidth={10}
      />
      <WithStyles(PolComponent)
        correct={false}
        hotspotColor="rgba(137, 183, 244, 0.65)"
        id="5"
        isDrawing={false}
        key="5"
        onClick={[Function]}
        onDragEnd={[Function]}
        outlineColor="blue"
        points={
          Array [
            Object {
              "x": 279,
              "y": 150,
            },
            Object {
              "x": 279,
              "y": 289,
            },
            Object {
              "x": 407,
              "y": 289,
            },
            Object {
              "x": 407,
              "y": 150,
            },
          ]
        }
        strokeWidth={10}
      />
    </Layer>
  </ForwardRef>
</div>
`;
